force-app/main/default/classes/generateCertificates.cls

```diff
@@ -0,0 +1,43 @@
+public class generateCertificates {    
+    @InvocableMethod(label='Generate certificates' description='Generate certificates for every participant of the training')
+    public static void generatePdfs(List<Id> ids) {
+        Id trainingId = ids[0];
+        List<Participant__c> participants = [SELECT Id, Training__c, Contact_Name__r.Name
+                                            FROM Participant__c
+                                            WHERE Training__c = :trainingId];
+
+        List<ContentVersion> listcv = new List<ContentVersion>();
+        List<ContentDocumentLink> listcdl = new List<ContentDocumentLink>();
+            
+        for(Participant__c particip :participants) {
+        	PageReference pageref = page.certificate;
+        	pageref.getParameters().put('Id', particip.Id); 
+        	pageref.setRedirect(true);
+        	Blob bdata = pageref.getContent();
+        
+        	ContentVersion cv = new ContentVersion();
+        	cv.ContentLocation = 'S';
+        	cv.PathOnClient = particip.Contact_Name__r.Name + '.pdf';
+        	cv.Title = particip.Contact_Name__r.Name + '.pdf';
+        	cv.VersionData = bdata;
+                
+            listcv.add(cv);
+        }
+        insert listcv;
```

Comment:
You can create files without creating ContentDocumentLinks. Here is an example from one of my projects.

`public static ContentVersion createDocumentVersion(Id contractId) {
    PageReference pdfPage = Page.ContractPDFFromContract;
    pdfPage.getParameters().put('id', contractId);

    Blob content;
    if(Test.isRunningTest()) {
        content = Blob.valueOf('TestContract');
    } else if(pdfPage == null) {
        content = Blob.valueOf(''); 
    } else {
        content = pdfPage.getContent();
    }

    ContentVersion version = new ContentVersion();
    version.ContentLocation = 'S'; 
    version.Title = DOCUMENT_NAME_DEFAULT;
    version.PathOnClient = DOCUMENT_NAME_DEFAULT;
    version.VersionData = content;
    version.FirstPublishLocationId = contractId;    
    
    return version; // you still need to insert it somewhere
}`